homeassistant/components/android_ip_webcam.py

```diff
@@ -0,0 +1,317 @@
+    'ffc': 'mdi:camera-front-variant',
+    'focus': 'mdi:image-filter-center-focus',
+    'gps_active': 'mdi:crosshairs-gps',
+    'light': 'mdi:flashlight',
+    'motion': 'mdi:run',
+    'night_vision': 'mdi:weather-night',
+    'overlay': 'mdi:monitor',
+    'pressure': 'mdi:gauge',
+    'proximity': 'mdi:map-marker-radius',
+    'quality': 'mdi:quality-high',
+    'sound': 'mdi:speaker',
+    'sound_event': 'mdi:speaker',
+    'sound_timeout': 'mdi:speaker',
+    'torch': 'mdi:white-balance-sunny',
+    'video_chunk_len': 'mdi:video',
+    'video_connections': 'mdi:eye',
+    'whitebalance_lock': 'mdi:white-balance-auto'
+}
+
+SWITCHES = ['audio_only', 'exposure_lock', 'ffc', 'focus',
+            'focus_homing', 'gps_active', 'idle', 'ivideon_streaming',
+            'motion_detect', 'night_vision', 'overlay', 'torch',
+            'whitebalance_lock']
+
+SENSORS = ['audio_connections', 'battery_level', 'battery_temp',
+           'battery_voltage', 'light', 'motion', 'pressure', 'proximity',
+           'sound', 'video_connections']
+
+CONF_MOTION_BINARY_SENSOR = 'motion_binary_sensor'
+
+DEFAULT_MOTION_BINARY_SENSOR = True
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Required(CONF_HOST): cv.string,
+        vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+        vol.Inclusive(CONF_USERNAME, 'authentication'): cv.string,
+        vol.Inclusive(CONF_PASSWORD, 'authentication'): cv.string,
+        vol.Optional(CONF_SWITCHES,
+                     default=SWITCHES): vol.All(cv.ensure_list,
+                                                [vol.In(SWITCHES)]),
+        vol.Optional(CONF_SENSORS,
+                     default=SENSORS): vol.All(cv.ensure_list,
+                                               [vol.In(SENSORS)]),
+        vol.Optional(CONF_MOTION_BINARY_SENSOR,
+                     default=DEFAULT_MOTION_BINARY_SENSOR): bool
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+ALLOWED_ORIENTATIONS = ['landscape', 'upsidedown', 'portrait',
+                        'upsidedown_portrait']
+
+MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=3)
+
+
+def setup(hass, config):
+    """Setup the IP Webcam component."""
+    conf = config[DOMAIN]
+    host = conf[CONF_HOST]
+    ip_webcam = hass.data.get(DATA_IP_WEBCAM)
+    if ip_webcam is None:
+        hass.data[DATA_IP_WEBCAM] = {}
+    hass.data[DATA_IP_WEBCAM][host] = IPWebcam(conf)
+
+    if conf.get(CONF_MOTION_BINARY_SENSOR, False) is True:
+        discovery.load_platform(hass, 'binary_sensor', DOMAIN, {}, config)
+
+    discovery.load_platform(hass, 'camera', DOMAIN, {}, config)
+
+    sensor_config = conf.get(CONF_SENSORS, [])
+    discovery.load_platform(hass, 'sensor', DOMAIN, sensor_config, config)
+
+    switch_config = conf.get(CONF_SWITCHES, [])
+    discovery.load_platform(hass, 'switch', DOMAIN, switch_config, config)
+
+    return True
+
+
+class IPWebcam(object):
```

Comment:
This is the kind of thing that is perfect to be extracted into a third party lib.